[WIP] Use a Form as the SDL window

flibitijibibo opened the pull request:
Why fake a Form.Handle pointer when we can use a real one!

This adds some stuff to create a Form that's also capable of handling the FNA OpenGL context. Amazingly it works great except for some issues with the viewport and window size. I'm not 100% sure why this is the case, but I have very little experience with Forms.

My test case is Stardew Valley; with these changes the game should boot and run and you'll even get working text input, but for some odd reason the viewport is stuck in the window's original size. If you call SDL_ShowWindow right after creating the SDL2 pointer it'll be correct until you do a style or size change, then it freaks out and goes back to a weird viewport.

AFAIK the viewport size is the only known issue; once we figure out why that's broken this should be good to go. EDIT: It also appears that ShowCursor doesn't quite work... kinda weird since that's not dependent on an SDL_Window, might be a deeper issue there.

madewokherd commented:
I'm curious whether SDL creates a child HWND, or attaches to the Form's HWND. Depending on which one, I expect we'd need to sync up the Form HWND's properties with the SDL HWND, or suppress winforms' default message handling for things like drawing.

And since we'll want to replace Mono's winforms implementation with .NET core, which I (unfortunately) know works very differently, it's probably not a good idea to put too much work into this just yet.

flibitijibibo commented:
Thankfully CreateWindowFrom is very boring:

https://hg.libsdl.org/SDL/file/0ff5bbe35bf5/src/video/SDL_video.c#l1572
https://hg.libsdl.org/SDL/file/0ff5bbe35bf5/src/video/windows/SDL_windowswindow.c#l377

The only scary part is SDL_HINT_VIDEO_WINDOW_SHARE_PIXEL_FORMAT, which might possibly be the worst way to inject an OpenGL format into a foreign Win32 window, but it worked...

madewokherd commented:
I see, so it overrides the window proc. That might not be good.

madewokherd commented:
Well, at least I can't think of any specific thing an application might do with winforms that we couldn't support this way.

madewokherd commented:
BTW, the ShowCursor issue is probably because their window proc doesn't handle WM_SETCURSOR when it's overriding (confusingly aka subclassing) another window proc: https://hg.libsdl.org/SDL/file/0ff5bbe35bf5/src/video/windows/SDL_windowsevents.c#l852

This approach requires us to handle cases like that.

madewokherd commented:
Judging by their window procedure, we'll also want special handling for
WM_GETMINMAXINFO
WM_WINDOWPOSCHANGING
WM_PAINT
WM_DROPFILES

To make a confusing term even more confusing, we may also be required to counter-subclass the window so we can override some of SDL's behaviors.

flibitijibibo commented:
Double checked to be 100% sure: The cursor is correctly invisible on startup, until you move the cursor, at which point it becomes visible. Going to take a guess and say we have to sync the Cursor.Hide/Show value, possibly in addition to the Size value. EDIT: Yup, that fixed the mouse. Just need to deal with this bizarre window size behavior...

madewokherd commented:
Might be better to subclass (in the normal OOP sense) Form and short-circuit winforms' handling of those messages in WndProc.

flibitijibibo committed:
Apply SynchronizeWithVerticalRetrace before OnPreparingDeviceSettings

flibitijibibo committed:
Move a bunch of SDL_GL_* calls to an isolated function

flibitijibibo committed:
Pretty things up for future GLDevices...

TheSpydog committed:
Moved SetPresentationInterval into GLDevices

flibitijibibo committed:
GraphicsDevice Resets should update the PresentationInterval

flibitijibibo committed:
Replace direct SDL call with FNAPlatform call

flibitijibibo commented:
Been playing with various tests and it seems like the Forms stuff as it is right now seems perfectly fine...? It's weird, all my tests are totally fine but Stardew's viewport goes nuts and I'm not sure why yet.

I've added a HackForm that is largely based on the XnaToFna Form implementation; my thought is that Stardew might be using some stuff like WindowState and FormBorderStyle so the easiest fix might simply be to do what we were doing before the ABI stuff:

https://github.com/FNA-XNA/FNA/blob/wine-fna/src/FNAPlatform/SDL2_FNAPlatform.cs#L52

As you can see I had to use the `new` keyword for those properties, which I think ultimately makes these useless because they never get tripped... other than that, AFAIK these are known to work when we run SV through XnaToFna. @0x0ade might have a better idea.

madewokherd commented:
I'm sure there's an event that'll tell you when those properties change.

madewokherd commented:
Well, I'm wrong, I see no event we can use. Of course, we control winforms, so if we absolutely have to drop a hack in there we can.

There may be a way to do what you need with window message handling, but I'm out of time to look into that for today.

flibitijibibo commented:
Depending on how long we plan to keep Mono WinForms around there's also a secret mechanism we can use:

https://github.com/mono/mono/blob/master/mcs/class/System.Windows.Forms/System.Windows.Forms/XplatUI.cs#L99
https://github.com/mono/mono/blob/master/mcs/class/System.Windows.Forms/System.Windows.Forms/XplatUIDriver.cs

madewokherd commented:
So, I don't understand what this code is supposed to do and why, which makes it hard to figure out how we should do it. Why can't we set the window styles the way the application wants to?

flibitijibibo commented:
Style changes force a whole update of the Control's HWND, which includes POSCHANGED events which overrides what the window size is supposed to be. Open up Stardew Valley with this patch applied and pay close attention to the game's ClientSizeChanged log entries. The window gets _really_ messed up and AFAIK they only touch style, nothing else.

madewokherd commented:
Well, I'm sure it doesn't get really messed up on Windows when that happens, why does for us?

madewokherd commented:
Window_ClientSizeChanged(); Window.ClientBounds={X:0 Y:30 Width:1680 Height:1020}

That's just my monitor size minus the top panel. What's messed up about it?

madewokherd commented:
That's on dotnet core winforms btw, mono's winforms may do something different and probably broken.

flibitijibibo commented:
Interesting, can you verify that it breaks with the old Forms? If so then I guess we can skip this, but the main thing I'm seeing is viewport size breakages in addition to issues with Alt-Enter and general user resizing.

madewokherd commented:
With mono winforms, it seems to keep a window size of 1280x720 all the time, even if I click on the icon in the upper-right which I assume is a fullscreen toggle. That icon just toggles the border.

With dotnet core winforms, it's a constant flickering mess that is the size of my work area.

flibitijibibo commented:
@0x0ade might be able to correct me on this, but I believe the way Stardew's window options work is by setting FormBorderStyle for borderless toggling, and setting WindowState to Maximixed for fullscreen. It probably does both at the same time for fullscreen, instead of just setting GraphicsDeviceManager.IsFullScreen to true.

madewokherd commented:
Interesting. It should be sizing itself to the whole screen and not just the work area, right? I'm pretty sure maximization covers the screen, but only with some window styles.

flibitijibibo commented:
I thought the same thing, seems it's a problem even on Windows so I wouldn't fuss too much about the work area part:

https://steamcommunity.com/app/413150/discussions/0/2561864094342571511/
https://steamcommunity.com/app/413150/discussions/search/?q=fullscreen&gidforum=487876568238453267&include_deleted=1

flibitijibibo committed:
MojoShader already handles this

flibitijibibo committed:
MojoShader update

flibitijibibo committed:
Implement PipelineCache for Effect state changes (#245)

Hash calculation by @TheSpydog

flibitijibibo commented:
Latest push is a rebase against master, so it has a bunch of other stuff in it; all we care about for this is the SDL2_FNAPlatform.cs changes.

flibitijibibo committed:
Purge as much msvcrt as possible

flibitijibibo committed:
Theorafile update

flibitijibibo committed:
iOS should no longer reference msvcrt

flibitijibibo committed:
MojoShader update

flibitijibibo committed:
FAudio 19.05

flibitijibibo committed:
19.05

flibitijibibo committed:
glInvalidateFramebuffer for improved target swap performance on iOS

flibitijibibo committed:
Parens were in the wrong spot...

flibitijibibo committed:
Even more bad parens, maybe hit Ctrl+S next time flibit

flibitijibibo committed:
Invalidate draw framebuffers, including backbuffer

flibitijibibo committed:
Only invalidate special read buffers

flibitijibibo committed:
Extra spacing from previous glInvalidateFramebuffer commits

flibitijibibo committed:
i->j typo fix

flibitijibibo committed:
Make glTexEnvi optional for ES3

flibitijibibo committed:
Accept numSamplers value without capping it, fixes vertex texture support

flibitijibibo committed:
Submodule update

flibitijibibo committed:
19.06

foreverWIP committed:
Updated Intellisense descriptions

The description for DPadUp referred to down, and vice versa for DPadDown. Swapped the two descriptions to correct them

kg committed:
Fix colorconverter incorrectly using floats in some places

flibitijibibo committed:
Return the Win32 HWND for games that request the Handle

flibitijibibo committed:
SDL2# update

flibitijibibo committed:
FAudio 19.07

flibitijibibo committed:
19.07

flibitijibibo committed:
When cloning effects, copy the texture values from the source parameter list

flibitijibibo committed:
MojoShader update

Commit:
Add missing generic type forwards to XNA ABI-compatibility libraries.

kg committed:
Replace Queue<int> for sampler modification tracking with bool[]

flibitijibibo committed:
Extra space, unused usings

flibitijibibo committed:
More unused usings

flibitijibibo committed:
Yet another unused using

flibitijibibo committed:
One more unused using... for now

flibitijibibo committed:
Play should resume if the instance is paused

flibitijibibo commented:
Closing for now, will have to try this again later.
